### PR TITLE
Stats

### PR DESCRIPTION
**Pull request by murcake**

### CODE REVIEW

**Comment by murcake** [comment]
The initial conversation about this PR has started in #145. We've moved it because that PR was submitted from a fork and it wasn't convenient to review

File: warp/logging/generateMarkdown.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import json
```
Comment (murcake) [inline warp/logging/generateMarkdown.py:1]: The general code style for python code is to write function names and variables in "snake_case", as opposed to "camelCase"

File: warp/logging/generateMarkdown.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        size /= 1024.0
+    return f"{size:.{decimal_places}f}{unit}"
+
+
+def sizeOfFile(contractName, file, fileName):
+    jsonPath = os.path.abspath(
+        os.path.join(__file__, "../../../benchmark/", fileName + ".json")
+    )
+    if os.path.exists(jsonPath):
+        functionSteps = json.load(open(jsonPath, "r"))
```
Comment (murcake) [inline warp/logging/generateMarkdown.py]: There is no guarantee that the file will be closed eventually, the information written to the filesystem and the file descriptor returned to the OS, unless `close` is called on it. It is less terse, but we'll have to use `with` with all `open` functions

File: warp/logging/generateMarkdown.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    else:
+        functionSteps = {}
+
+    if contractName not in functionSteps.keys():
+        functionSteps[contractName] = {}
+    functionSteps[contractName][
+        f"Steps in {functionName}"
+    ] = result.call_info.cairo_usage.n_steps
```
[the remark this answers is not in the document]
Comment (murcake) [reply warp/logging/generateMarkdown.py:72]: The same can also be used above

File: warp/logging/generateMarkdown.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+
+def sizeOfFile(contractName, file, fileName):
+    jsonPath = os.path.abspath(
+        os.path.join(__file__, "../../../benchmark/", fileName + ".json")
+    )
+    if os.path.exists(jsonPath):
+        jsonFile = open(jsonPath, "r")
+        functionSteps = json.load(jsonFile)
+        jsonFile.close()
```
Comment (murcake) [inline warp/logging/generateMarkdown.py:21]: Context managers work better, you can never predict an exception:
```
wtih open(jsonPath, "r") as jsonFile:
  functionSteps = json.load(jsonFile)
```

File: warp/logging/generateMarkdown.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        sys.getsizeof(bytecode)
+    )
+    functionSteps[contractName]["Bytecode length"] = len(bytecode)
+
+    jsonFile = open(jsonPath, "w")
+    json.dump(functionSteps, jsonFile, indent=3)
+    jsonFile.close()
+
+
+def stepsInFunction(contractName, functionName, result, fileName):
```
Comment (murcake) [inline warp/logging/generateMarkdown.py:57]: Can you please add a type annotation for `result`?

File: .github/workflows/develop-ci.yml [hunk 6]
```diff
@@ -53,3 +53,19 @@ jobs:
 
       - name: Test yul tests
         run: make test_yul
+
+      - name: Generate benchmarks for tests
+        run: |
+          mkdir -p ./benchmark/stats
+          make test_bench
```
Comment (murcake) [inline .github/workflows/develop-ci.yml:60]: I wouldn't call it `test_bench`, we don't test anything, we just benchmark, but it doesn't matter all that much

**Review by murcake — COMMENTED** [review]
I'm not sure how it's going to work. Each CI run commits stat-information to a repo. What can we do with it then?

A realistic tool in my imagination given a commit hash `h`, plots a graph of all the stats changes in the range `[..h]`. 
It might be possible if we store a commit hash along with each stats file.

**Review by Glitch18 — APPROVED** [review]

**Pull request merged**